File: src/stripe_entitlements/price_policy.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any


def _currency_options_match(
    value: Any, *, expected_currency: str, expected_unit_amount: int
) -> bool:
    if value is None or value == {}:
        return True
    if not isinstance(value, Mapping) or set(value) != {expected_currency}:
        return False
    option = value.get(expected_currency)
    if not isinstance(option, Mapping):
        return False
    unit_amount = option.get("unit_amount")
    return bool(
        type(unit_amount) is int
        and unit_amount == expected_unit_amount
        and option.get("custom_unit_amount") is None
        and option.get("tax_behavior") in {None, "unspecified"}
    )


def _catalog_identity_matches(
    price: Mapping[str, Any],
    product: Mapping[str, Any],
    *,
    expected_product_line: str,
    expected_plan_key: str,
) -> bool:
    metadata_values = (price.get("metadata"), product.get("metadata"))
    for metadata in metadata_values:
        if metadata is None:
            continue
        if not isinstance(metadata, Mapping):
            return False
        observed_product_line = metadata.get("product_line")
        observed_plan = metadata.get("plan")
        if observed_product_line is not None and observed_product_line != expected_product_line:
            return False
        if observed_plan is not None and observed_plan != expected_plan_key:
            return False
    return True


def catalog_price_matches(
    price: Mapping[str, Any],
    *,
    expected_currency: str,
    expected_unit_amount: int,
    expected_interval: str,
    expected_product_line: str,
    expected_plan_key: str,
    expected_lookup_key: str | None = None,
    expected_price_id: str | None = None,
    require_active: bool = True,
) -> bool:
    recurring = price.get("recurring")
    product = price.get("product")
    price_active = price.get("active", True)
    product_active = product.get("active", True) if isinstance(product, Mapping) else None
    active_matches = bool(not require_active or (price_active is True and product_active is True))
    currency = price.get("currency")
    unit_amount = price.get("unit_amount")
    interval_count = recurring.get("interval_count", 1) if isinstance(recurring, Mapping) else None
    return bool(
        isinstance(recurring, Mapping)
        and isinstance(product, Mapping)
        and active_matches
        and _catalog_identity_matches(
            price,
            product,
            expected_product_line=expected_product_line,
            expected_plan_key=expected_plan_key,
        )
        and (expected_lookup_key is None or price.get("lookup_key") == expected_lookup_key)
        and (expected_price_id is None or price.get("id") == expected_price_id)
        and isinstance(currency, str)
        and currency == currency.lower()
        and currency == expected_currency
        and type(unit_amount) is int
        and unit_amount == expected_unit_amount
        and recurring.get("interval") == expected_interval
        and type(interval_count) is int
        and interval_count == 1
        and recurring.get("usage_type", "licensed") == "licensed"
        and price.get("type", "recurring") == "recurring"
        and price.get("billing_scheme", "per_unit") == "per_unit"
        and price.get("tiers_mode") is None
        and price.get("transform_quantity") is None
        and price.get("custom_unit_amount") is None
        and _currency_options_match(
            price.get("currency_options"),
            expected_currency=expected_currency,
            expected_unit_amount=expected_unit_amount,
        )
        and price.get("tax_behavior") in {None, "unspecified"}
    )

File: typescript/src/price-policy.ts
import type { BillingInterval } from "./types.js";

function isSupportedRecord(
  value: unknown,
): value is Readonly<Record<string, unknown>> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  try {
    const prototype: unknown = Object.getPrototypeOf(value);
    if (prototype !== null && prototype !== Object.prototype) {
      return false;
    }
    return Reflect.ownKeys(value).every((key) => {
      const descriptor = Object.getOwnPropertyDescriptor(value, key);
      return (
        typeof key === "string" &&
        descriptor !== undefined &&
        descriptor.enumerable &&
        "value" in descriptor
      );
    });
  } catch {
    return false;
  }
}

function isAbsent(value: unknown): boolean {
  return value === undefined || value === null;
}

function ownValueOrDefault(
  value: Readonly<Record<string, unknown>>,
  field: string,
  fallback: unknown,
): unknown {
  return Object.hasOwn(value, field) ? value[field] : fallback;
}

function currencyOptionsMatch(
  value: unknown,
  expectedCurrency: string,
  expectedUnitAmount: number,
): boolean {
  if (isAbsent(value)) {
    return true;
  }
  if (!isSupportedRecord(value)) {
    return false;
  }
  const keys = Object.keys(value);
  if (keys.length === 0) {
    return true;
  }
  if (keys.length !== 1 || keys[0] !== expectedCurrency) {
    return false;
  }
  const option = value[expectedCurrency];
  return (
    isSupportedRecord(option) &&
    typeof option["unit_amount"] === "number" &&
    Number.isSafeInteger(option["unit_amount"]) &&
    option["unit_amount"] === expectedUnitAmount &&
    isAbsent(option["custom_unit_amount"]) &&
    (isAbsent(option["tax_behavior"]) ||
      option["tax_behavior"] === "unspecified")
  );
}

function catalogIdentityMatches(
  price: Readonly<Record<string, unknown>>,
  product: Readonly<Record<string, unknown>>,
  expectedProductLine: string,
  expectedPlanKey: string,
): boolean {
  for (const metadata of [price["metadata"], product["metadata"]]) {
    if (isAbsent(metadata)) {
      continue;
    }
    if (!isSupportedRecord(metadata)) {
      return false;
    }
    const productLine = metadata["product_line"];
    const plan = metadata["plan"];
    if (!isAbsent(productLine) && productLine !== expectedProductLine) {
      return false;
    }
    if (!isAbsent(plan) && plan !== expectedPlanKey) {
      return false;
    }
  }
  return true;
}

export interface CatalogPriceExpectation {
  readonly expectedCurrency: string;
  readonly expectedUnitAmount: number;
  readonly expectedInterval: BillingInterval;
  readonly expectedProductLine: string;
  readonly expectedPlanKey: string;
  readonly expectedLookupKey?: string | null;
  readonly expectedPriceId?: string | null;
  readonly requireActive?: boolean;
}

/** Validate an expanded recurring Stripe Price against the immutable catalog. */
export function catalogPriceMatches(
  price: unknown,
  expected: CatalogPriceExpectation,
): boolean {
  if (
    !isSupportedRecord(price) ||
    !Number.isSafeInteger(expected.expectedUnitAmount)
  ) {
    return false;
  }
  const recurring = price["recurring"];
  const product = price["product"];
  if (!isSupportedRecord(recurring) || !isSupportedRecord(product)) {
    return false;
  }

  const requireActive = expected.requireActive ?? true;
  if (typeof requireActive !== "boolean") {
    return false;
  }
  const priceActive = ownValueOrDefault(price, "active", true);
  const productActive = ownValueOrDefault(product, "active", true);
  if (requireActive && (priceActive !== true || productActive !== true)) {
    return false;
  }

  const currency = price["currency"];
  const unitAmount = price["unit_amount"];
  const intervalCount = ownValueOrDefault(recurring, "interval_count", 1);
  return (
    catalogIdentityMatches(
      price,
      product,
      expected.expectedProductLine,
      expected.expectedPlanKey,
    ) &&
    (isAbsent(expected.expectedLookupKey) ||
      price["lookup_key"] === expected.expectedLookupKey) &&
    (isAbsent(expected.expectedPriceId) ||
      price["id"] === expected.expectedPriceId) &&
    typeof currency === "string" &&
    currency === currency.toLowerCase() &&
    currency === expected.expectedCurrency &&
    typeof unitAmount === "number" &&
    Number.isSafeInteger(unitAmount) &&
    unitAmount === expected.expectedUnitAmount &&
    recurring["interval"] === expected.expectedInterval &&
    typeof intervalCount === "number" &&
    Number.isSafeInteger(intervalCount) &&
    intervalCount === 1 &&
    ownValueOrDefault(recurring, "usage_type", "licensed") === "licensed" &&
    ownValueOrDefault(price, "type", "recurring") === "recurring" &&
    ownValueOrDefault(price, "billing_scheme", "per_unit") === "per_unit" &&
    isAbsent(price["tiers_mode"]) &&
    isAbsent(price["transform_quantity"]) &&
    isAbsent(price["custom_unit_amount"]) &&
    currencyOptionsMatch(
      price["currency_options"],
      expected.expectedCurrency,
      expected.expectedUnitAmount,
    ) &&
    (isAbsent(price["tax_behavior"]) || price["tax_behavior"] === "unspecified")
  );
}

export interface CatalogOneTimePriceExpectation {
  readonly expectedCurrency: string;
  readonly expectedUnitAmount: number;
  readonly expectedProductLine: string;
  readonly expectedPackKey: string;
  readonly expectedLookupKey: string;
  readonly requireActive?: boolean;
}

/** Validate the deliberately narrow one-time credit-pack Stripe Price shape. */
export function catalogOneTimePriceMatches(
  price: unknown,
  expected: CatalogOneTimePriceExpectation,
): boolean {
  if (
    !isSupportedRecord(price) ||
    !Number.isSafeInteger(expected.expectedUnitAmount)
  ) {
    return false;
  }
  const product = price["product"];
  if (!isSupportedRecord(product)) {
    return false;
  }
  for (const metadata of [price["metadata"], product["metadata"]]) {
    if (isAbsent(metadata)) {
      continue;
    }
    if (
      !isSupportedRecord(metadata) ||
      (!isAbsent(metadata["product_line"]) &&
        metadata["product_line"] !== expected.expectedProductLine) ||
      (!isAbsent(metadata["credit_pack"]) &&
        metadata["credit_pack"] !== expected.expectedPackKey) ||
      !isAbsent(metadata["plan"])
    ) {
      return false;
    }
  }

  const requireActive = expected.requireActive ?? true;
  if (typeof requireActive !== "boolean") {
    return false;
  }
  const currency = price["currency"];
  const unitAmount = price["unit_amount"];
  return (
    (!requireActive ||
      (price["active"] === true && product["active"] === true)) &&
    price["lookup_key"] === expected.expectedLookupKey &&
    ownValueOrDefault(price, "type", "one_time") === "one_time" &&
    isAbsent(price["recurring"]) &&
    typeof currency === "string" &&
    currency === currency.toLowerCase() &&
    currency === expected.expectedCurrency &&
    typeof unitAmount === "number" &&
    Number.isSafeInteger(unitAmount) &&
    unitAmount === expected.expectedUnitAmount &&
    ownValueOrDefault(price, "billing_scheme", "per_unit") === "per_unit" &&
    isAbsent(price["tiers_mode"]) &&
    isAbsent(price["transform_quantity"]) &&
    isAbsent(price["custom_unit_amount"]) &&
    (isAbsent(price["tax_behavior"]) ||
      price["tax_behavior"] === "unspecified") &&
    currencyOptionsMatch(
      price["currency_options"],
      expected.expectedCurrency,
      expected.expectedUnitAmount,
    )
  );
}
